### ISSUE TITLE
UndefinedFunction.diff broken in sympy 1.9, 1.10rc2

### ISSUE THREAD

**Issue by mkoeppe** [sympy/sympy#23144]
```
>>> import sympy
>>> sympy.__version__
'1.10rc2'
>>> from sympy.abc import x, y
>>> from sympy import Function
>>> f = Function('f')
>>> f.diff(x)
1
>>> f.diff(x,x)
1
>>> f.diff(x, 2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mkoeppe/s/sage/sage-rebasing/local/var/lib/sage/venv-python3.9/lib/python3.9/site-packages/sympy/core/expr.py", line 3526, in diff
    return _derivative_dispatch(self, *symbols, **assumptions)
  File "/Users/mkoeppe/s/sage/sage-rebasing/local/var/lib/sage/venv-python3.9/lib/python3.9/site-packages/sympy/core/function.py", line 1919, in _derivative_dispatch
    return Derivative(expr, *variables, **kwargs)
  File "/Users/mkoeppe/s/sage/sage-rebasing/local/var/lib/sage/venv-python3.9/lib/python3.9/site-packages/sympy/core/function.py", line 1287, in __new__
    raise ValueError("First variable cannot be a number: %i" % v)
ValueError: First variable cannot be a number: 2
>>> f.diff(x,2,y,2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mkoeppe/s/sage/sage-rebasing/local/var/lib/sage/venv-python3.9/lib/python3.9/site-packages/sympy/core/expr.py", line 3526, in diff
    return _derivative_dispatch(self, *symbols, **assumptions)
  File "/Users/mkoeppe/s/sage/sage-rebasing/local/var/lib/sage/venv-python3.9/lib/python3.9/site-packages/sympy/core/function.py", line 1919, in _derivative_dispatch
    return Derivative(expr, *variables, **kwargs)
  File "/Users/mkoeppe/s/sage/sage-rebasing/local/var/lib/sage/venv-python3.9/lib/python3.9/site-packages/sympy/core/function.py", line 1287, in __new__
    raise ValueError("First variable cannot be a number: %i" % v)
ValueError: First variable cannot be a number: 2
```

(seen in https://github.com/sympy/sympy/runs/4929630109?check_suite_focus=true, failure for `src/sage/symbolic/expression_conversions.py`)

**Comment by oscarbenjamin** [comment]
Thanks for picking up on this. I'll take a look.

I didn't actually realise that it was possible to use a Function like this at all. Is this used somewhere in Sage?

**Comment by oscarbenjamin** [comment]
I think that there is a confusion here. Ideally Functions would be a proper objects but as it stands they are not (#4787).

The object `f` returned by `Function('f')` is actually a Python class. Then calling `f(x)` creates an instance of that class which is then a proper SymPy expression instance (a subclass of Basic/Expr):
```python
In [1]: f = Function("f")

In [2]: type(f)
Out[2]: sympy.core.function.UndefinedFunction

In [3]: f = Function("f")

In [4]: f.mro()
Out[4]: 
[f,
 AppliedUndef,
 Function,
 Application,
 sympy.core.expr.Expr,
 sympy.core.basic.Basic,
 sympy.printing.defaults.Printable,
 sympy.core.evalf.EvalfMixin,
 object]

In [5]: fx = f(x)

In [6]: type(fx)
Out[6]: f

In [7]: isinstance(fx, f)
Out[7]: True
```
When you call `f.diff(x, x)` the first `x` is being passed as self to the bound class method:
```python
In [8]: f.diff
Out[8]: <function sympy.core.expr.Expr.diff(self, *symbols, **assumptions)>

In [9]: f(x).diff
Out[9]: <bound method Expr.diff of f(x)>

In [10]: f.diff(x)
Out[10]: 1

In [11]: x.diff()
Out[11]: 1
```
When you pass a second argument that is treated as the argument:
```python
In [12]: f.diff(x ** 2, x)
Out[12]: 2⋅x
```
Calling `f.diff(x, 2)` means differentiating wrt 2:
```python
In [13]: f.diff(x, 2)
...
ValueError: First variable cannot be a number: 2

In [14]: x.diff(2)
...
ValueError: First variable cannot be a number: 2
```
If this is a new error showing up in the Sage tests then there must be something else that is the cause because this behaviour for `f.diff(x, 2)` is seen as far back as SymPy 1.2.

**Comment by oscarbenjamin** [comment]
I looked at the test here:
https://github.com/sagemath/sage/blob/15c8011dd0b21c74f698b05ed2608f7afbec0d1a/src/sage/symbolic/expression_conversions.py#L863-L871
It's doing:
```
            sage: f_sage = function('f_sage')(x, y)
            sage: f_sympy = f_sage._sympy_()
```
So I would guess that `f_sympy` is supposed to be equivalent to `f(x, y)` like:
```python
In [15]: f = Function("f")

In [16]: x, y = symbols("x, y")

In [17]: f(x, y).diff(x, 2)
Out[17]: 
  2         
 ∂          
───(f(x, y))
  2         
∂x 
```

**Comment by mkoeppe** [comment]
I agree that the Sage doctest that is triggering this is meaningless and should be changed.

Nevertheless, ideally `f.diff(...)` and `diff(f, ...)` should probably give a meaningful error message in sympy.
The same is, by the way, also happening for defined functions:
```
>>> from sympy import sin
>>> from sympy.abc import x
>>> sin.diff(x)
1
```

**Comment by mkoeppe** [comment]
That we are now seeing this with SymPy 1.10rc2 in the Sage testsuite seems to actually come from a change how Sage integers are handled somewhere in this codepath. 
```
sage: import sympy
sage: sympy.__version__
'1.10rc2'
sage: from sympy.abc import x
sage: x.diff(2)
ValueError: 
Can't calculate derivative wrt 2.
```

With 1.9, we got:
```
sage: import sympy
sage: sympy.__version__
'1.9'
sage: from sympy.abc import x
sage: x.diff(2)
ValueError: First variable cannot be a number: 2
```

**Comment by mkoeppe** [comment]
This is from https://github.com/sympy/sympy/commit/626064e1a582d6dc37fc3bcf54e3508d38fe6877

**Comment by mkoeppe** [comment]
Ad-hoc hardcoded type lists like  `array_likes = (tuple, list, Tuple)`, `integer_likes = (int, Integer)` should probably be avoided in favor of standard ABCs

**Comment by oscarbenjamin** [comment]
It seems to be this line where it goes astray:
https://github.com/sympy/sympy/blob/bff5b2ede99b0ff18434bc8f22a4ee76b2acbb77/sympy/core/function.py#L1284
```
(Pdb) l
1280 	        integer_likes = (int, Integer)
1281 	
1282 	        from sympy.tensor.array import Array, NDimArray
1283 	
1284 	        for i, v in enumerate(variables):
1285 ->	            if isinstance(v, integer_likes):
1286 	                if i == 0:
1287 	                    raise ValueError("First variable cannot be a number: %i" % v)
1288 	                count = v
1289 	                prev, prevcount = variable_count[-1]
1290 	                if prevcount != 1:
(Pdb) p v
2
(Pdb) p type(v)
<class 'sage.rings.integer.Integer'>
(Pdb) p isinstance(v, integer_likes)
False
```
I think the problem is the fact that `v` is not sympified:
```
(Pdb) sympify(v)
2
(Pdb) type(sympify(v))
<class 'sympy.core.numbers.Integer'>
```
CC @ThePauliPrinciple

**Comment by mkoeppe** [comment]
> I looked at the test here: https://github.com/sagemath/sage/blob/15c8011dd0b21c74f698b05ed2608f7afbec0d1a/src/sage/symbolic/expression_conversions.py#L863-L871 It's doing:
> 
> ```
>             sage: f_sage = function('f_sage')(x, y)
>             sage: f_sympy = f_sage._sympy_()
> ```
> 
> So I would guess that `f_sympy` is supposed to be equivalent to `f(x, y)` 

You are right. The problem is really that Sage integers are no longer handled. This has nothing to do with `UndefinedFunction`.
```
sage: var('y')
y
sage: f_sage = function('f_sage')(x, y)
sage: f_sage
f_sage(x, y)
sage: f_sympy = f_sage._sympy_()
sage: f_sympy
f_sage(x, y)
sage: f_sympy.diff(x, 2)
ValueError: 
Can't calculate derivative wrt 2.
sage: f_sympy.diff(x, 2r)                 # Use 2 as Python int
Derivative(f_sage(x, y), (x, 2))
```

**Comment by mkoeppe** [comment]
> Ad-hoc hardcoded type lists like `array_likes = (tuple, list, Tuple)`, `integer_likes = (int, Integer)` should probably be avoided in favor of standard ABCs

Done in #23145

**Comment by mkoeppe** [comment]
`git grep 'int, Integer'` shows many more places that could be changed similarly; but perhaps these are taken care of `sympify`?

**Comment by oscarbenjamin** [comment]
Most of them are in the rubi module which can be ignored. I think a lot of the others should really just be `isinstance(obj, Integer)`. Sometimes contributors get confused about this and add paths to allow `int` in places where it shouldn't really be allowed because it should have been handled by `sympify`.

**Comment by mkoeppe** [comment]
I've now revised #23145 to use a test similar to one that I see in `sympy/core/numbers.py`

**Comment by mkoeppe** [comment]
`ArrayDerivative` is also affected:
```
sage:     from sympy.matrices.expressions.matexpr import MatrixSymbol
sage:     from sympy.tensor.array.array_derivatives import ArrayDerivative
....:     A = MatrixSymbol("A", 2, 2)
sage: arrder = ArrayDerivative(A, A, 2, evaluate=False)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-299c760395dd> in <module>
----> 1 arrder = ArrayDerivative(A, A, Integer(2), evaluate=False)

~/s/sage/sage-rebasing/worktree-gcc11/local/var/lib/sage/venv-python3.9/lib/python3.9/site-packages/sympy/tensor/array/array_derivatives.py in __new__(cls, expr, *variables, **kwargs)
     17 
     18     def __new__(cls, expr, *variables, **kwargs):
---> 19         obj = super().__new__(cls, expr, *variables, **kwargs)
     20         if isinstance(obj, ArrayDerivative):
     21             obj._shape = obj._get_shape()

~/s/sage/sage-rebasing/worktree-gcc11/local/var/lib/sage/venv-python3.9/lib/python3.9/site-packages/sympy/core/function.py in __new__(cls, expr, *variables, **kwargs)
   1343             if not v._diff_wrt:
   1344                 __ = ''  # filler to make error message neater
-> 1345                 raise ValueError(filldedent('''
   1346                     Can't calculate derivative wrt %s.%s''' % (v,
   1347                     __)))

ValueError: 
Can't calculate derivative wrt 2.
```

**Comment by mkoeppe** [comment]
Also `MatrixSymbol.__getitem__`
```
sage:     >>> from sympy import MatrixSymbol
....:     >>> A = MatrixSymbol('A', 3, 3)
....: 
sage: A[1]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-2-5d6865200369> in <module>
----> 1 A[Integer(1)]

~/s/sage/sympy/sympy/matrices/expressions/matexpr.py in __getitem__(self, key)
    311                 Only integers may be used when addressing the matrix
    312                 with a single index.'''))
--> 313         raise IndexError("Invalid index, wanted %s[i,j]" % self)
    314 
    315     def _is_shape_symbolic(self) -> bool:

IndexError: Invalid index, wanted A[i,j]
```

**Comment by oscarbenjamin** [comment]
I think these are all situations where sympify should be used somewhere. It's just fiddly to know exactly where to call sympify if you have something like an argument that might be a number or might be a tuple/list etc.

For example in the case of `Matrix.__getitem__` there should be cases to handle tuple/slice and then for the remaining cases the key should be sympified.

**Comment by mkoeppe** [comment]
I have fixes that seem to work for me in #23145; please feel free to edit to taste.

**Comment by asmeurer** [comment]
> Nevertheless, ideally f.diff(...) and diff(f, ...) should probably give a meaningful error message in sympy.
> The same is, by the way, also happening for defined functions:

I agree with this in principle, but what you are seeing is really just an issue with Python itself. If you have

```py
class A:
    def method(self, x):
        ...
```

Then `method` is intended to be called like `A().method(x)`. But Python also lets you call it like `A.method(A(), x)`, and indeed you can also call `A.method(B(), x)`:

```py
>>> class A:
...     def method(self, x):
...         print(self, x)
>>> A().method(1)
<__main__.A object at 0x10e666400> 1
>>> A.method(A(), 1)
<__main__.A object at 0x11ea4cc70> 1
>>> A.method(0, 1)
0 1
```

So what is happening here is you are calling the `diff` method on the `sin` class, which usually expects the first argument to be a `sin` instance, but instead passing it `Symbol('x')` (the second argument to diff here is not passed at all; it is optional in SymPy if the expression only has one symbol in it). It just so happens that `sin.diff` method dispatches to `diff` and so it takes the derivative of `x`. 

There's not really much we can do to avoid this gotcha. It's unfortunate that Python lets you do this (there are some good reasons for it), but it's just the way that it works. At best, we could make it so that things like `f` and `sin` aren't actually Python classes, so that they don't have this behavior. But this is a much harder change to make. As you can see, #4787 has been open for over 12 years.

**Comment by mkoeppe** [comment]
Thanks for the analysis!

**Comment by mkoeppe** [comment]
It'd be fine with me to close the present issue and continue on #23145 with the actual problem.

**Comment by ThePauliPrinciple** [comment]
Mind you, I have given up on trying to streamline sympification for `numbers`, because there are just too many things that need to be taken into account in combination with the somewhat inconsistent behaviour in SymPy (when do we allow builtin types and when do we not allow them etc.)

**Comment by oscarbenjamin** [comment]
The original problem is now fixed by #23162 on master and https://github.com/sympy/sympy/pull/23166 on the 1.10 release branch.

**Comment by mkoeppe** [comment]
Perfect!
